raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -109,50 +114,86 @@ private OptionalLong quorumMajorityFetchTimestamp() {
         // To be resilient to system time shifts we do not strictly require the timestamp be monotonically increasing
         state.lastFetchTimestamp = OptionalLong.of(Math.max(state.lastFetchTimestamp.orElse(-1L), timestamp));
         return quorumMajorityFetchTimestamp();
     }
 
     public List<Integer> nonLeaderVotersByDescendingFetchOffset() {
         return followersByDescendingFetchOffset().stream()
-                   .filter(state -> state.nodeId != localId)
-                   .map(state -> state.nodeId).collect(Collectors.toList());
+                   .filter(state -> state.replicaId() != localId)
+                   .map(ReplicaState::replicaId).collect(Collectors.toList());
     }
 
     private List<ReplicaState> followersByDescendingFetchOffset() {
-        List<ReplicaState> followersByDescendingFetchOffset = new ArrayList<>(this.voterReplicaStates.values());
-        Collections.sort(followersByDescendingFetchOffset);
-        return followersByDescendingFetchOffset;
+        return new ArrayList<>(this.voterReplicaStates.values())
+            .stream().sorted().map(voterState -> new ReplicaState()
+                .setReplicaId(voterState.nodeId)
+                .setLastCaughtUpTimeMs(voterState.lastFetchTimestamp.orElse(-1))
+                .setLogEndOffset(voterState.endOffset.orElse(-1)))
+                .collect(Collectors.toList());
     }
 
     /**
      * @return true if the high watermark is updated too
      */
     public boolean updateEndOffset(int remoteNodeId, long endOffset) {
-        ReplicaState state = ensureValidVoter(remoteNodeId);
+        NodeState state = getReplicaState(remoteNodeId);
+
         state.endOffset.ifPresent(currentEndOffset -> {
             if (currentEndOffset > endOffset)
                 throw new IllegalArgumentException("Non-monotonic update to end offset for nodeId " + remoteNodeId);
         });
-        state.hasEndorsedLeader = true;
         state.endOffset = OptionalLong.of(endOffset);
-        return updateHighWatermark();
+        if (isVoter(remoteNodeId)) {
+            addEndorsementFrom(remoteNodeId);
+            return updateHighWatermark();
+        }
+        return false;
     }
 
     public void addEndorsementFrom(int remoteNodeId) {
-        ReplicaState replicaState = ensureValidVoter(remoteNodeId);
-        replicaState.hasEndorsedLeader = true;
+        VoterState voterState = ensureValidVoter(remoteNodeId);
+        voterState.hasEndorsedLeader = true;
     }
 
-    private ReplicaState ensureValidVoter(int remoteNodeId) {
-        ReplicaState state = voterReplicaStates.get(remoteNodeId);
+    private VoterState ensureValidVoter(int remoteNodeId) {
+        VoterState state = voterReplicaStates.get(remoteNodeId);
         if (state == null)
             throw new IllegalArgumentException("Unexpected endorsement from non-voter " + remoteNodeId);
         return state;
     }
 
+    NodeState getReplicaState(int remoteNodeId) {
```

Comment:
This name will be clearer if we retain `ReplicaState`. Otherwise we should change it.